app/src/main/java/com/android/sample/ui/listing/ListingViewModel.kt

```
@@ -138,6 +143,29 @@ class ListingViewModel(
     }
   }
 
+  /**
+   * Check if the current user already has a booking for this listing
+   *
+   * @param listingId The ID of the listing
+   * @param userId The current user's ID
+   */
+  private fun checkExistingBooking(listingId: String, userId: String?) {
+    if (userId == null) return
+
+    viewModelScope.launch {
+      try {
+        val bookings = bookingRepo.getBookingsByListing(listingId)
+        val hasExistingBooking =
+            bookings.any { booking ->
+              booking.bookerId == userId && booking.status != BookingStatus.CANCELLED
+            }
+        _uiState.update { it.copy(hasExistingBooking = hasExistingBooking) }
+      } catch (e: Exception) {
+        Log.w("ListingViewModel", "Failed to check existing bookings", e)
```

Consider updating loading/error state here because currently if it fails, the UI will behave as if there is no existing booking, which could be a problematic silent failure.

---

done